nixos/modules/services/torrent/deluge.nix

```diff
@@ -15,50 +39,160 @@ in {
+            when <option>services.deluge.declarative</option> is set to
+            <literal>true</literal>.
+            See https://dev.deluge-torrent.org/wiki/UserGuide/Authentication for
+            more informations.
+          '';
+        };
       };
 
-      deluge.web.enable = mkEnableOption "Deluge Web daemon";
+      deluge.web = {
+        enable = mkEnableOption "Deluge Web daemon";
+        port = mkOption {
+        type = types.port;
+          default = 8112;
+          description = ''
+            Deluge web UI port.
+          '';
+        };
+      };
     };
   };
 
   config = mkIf cfg.enable {
 
+    systemd.tmpfiles.rules = [ "d '${configDir}' 0770 deluge deluge" ]
+    ++ optional (hasAttr "download_location" cfg.config)
+      "d '${cfg.config.download_location}' 0770 deluge deluge"
+    ++ optional (hasAttr "torrentfiles_location" cfg.config)
+      "d '${cfg.config.torrentfiles_location}' 0770 deluge deluge"
+    ++ optional (hasAttr "move_completed_path" cfg.config)
+      "d '${cfg.config.move_completed_path}' 0770 deluge deluge";
+
     systemd.services.deluged = {
       after = [ "network.target" ];
       description = "Deluge BitTorrent Daemon";
       wantedBy = [ "multi-user.target" ];
       path = [ pkgs.deluge ];
       serviceConfig = {
-        ExecStart = "${pkgs.deluge}/bin/deluged -d";
-        # To prevent "Quit & shutdown daemon" from working; we want systemd to manage it!
+        ExecStart = ''
+          ${pkgs.deluge}/bin/deluged \
+            --do-not-daemonize \
+            --config ${configDir}
+        '';
+        # To prevent "Quit & shutdown daemon" from working; we want systemd to
+        # manage it!
         Restart = "on-success";
         User = "deluge";
         Group = "deluge";
+        UMask = "0002";
         LimitNOFILE = cfg.openFilesLimit;
       };
+      preStart = preStart;
     };
 
     systemd.services.delugeweb = mkIf cfg_web.enable {
-      after = [ "network.target" ];
+      after = [ "network.target" "deluged.service"];
+      requires = [ "deluged.service" ];
       description = "Deluge BitTorrent WebUI";
       wantedBy = [ "multi-user.target" ];
       path = [ pkgs.deluge ];
-      serviceConfig.ExecStart = "${pkgs.deluge}/bin/deluge --ui web";
-      serviceConfig.User = "deluge";
-      serviceConfig.Group = "deluge";
+      serviceConfig = {
+        ExecStart = ''
+          ${pkgs.deluge}/bin/deluge-web \
+            --config ${configDir} \
+            --port ${toString cfg.web.port}
+        '';
+        User = "deluge";
+        Group = "deluge";
+      };
+    };
+
+    networking.firewall = mkIf (cfg.declarative && cfg.openFirewall) {
+      allowedTCPPortRanges = optional (hasAttr "listen_ports" cfg.config)
+        (listToRange cfg.config.listen_ports or listenPortsDefault);
+      allowedUDPPortRanges = optional (hasAttr "listen_ports" cfg.config)
+        (listToRange cfg.config.listen_ports or listenPortsDefault);
```

Comment:
This doesn't work like that, because it still only opens the ports when `listen_ports` is set in the config, and `listenPortsDefault` won't ever be used (in fact it's undefined right now). This is what should work:

```nix
{
  allowedTCPRanges = singleton (listToRange (cfg.config.listen_ports or [6881 6891]));
}
```

But now I'm also noticing that those ports won't even be used if `random_port` is enabled (which it is by default), so you should make this

```nix
{
  networking.firewall = mkIf (cfg.declarative && cfg.openFirewall && !(cfg.config.random_port or true)) {
    allowedTCPPortRanges = singleton (listToRange (cfg.config.listen_ports or [6881 6891]));
    allowedUDPPortRanges = singleton (listToRange (cfg.config.listen_ports or [6881 6891]));
  };
}
```

And now I'm debating whether it's worth even putting this in the module, because I'd think most people would want `openFirewall` to open the daemon or web port, which it doesn't even do now, and with the default settings it doesn't even open anything unless you turn off `random_port` (which most users will leave enabled). Does opening these ports even do anything for most people? I never opened these in my config and everything seems to be working just fine.

Reply:
> Does opening these ports even do anything for most people? I never opened these in my config and everything seems to be working just fine.

Even though you can connect to some pairs this is probably due to some additions to BitTorrent protocol. If you want to seed [`bep03`](http://www.bittorrent.org/beps/bep_0003.html) the original protocol specification state it clearly:
> Common behavior is for a downloader to try to listen on port 6881 and if that port is taken try 6882, then 6883, etc. and give up after 6889.